### PR TITLE
fix(fetch): remove terminated event listener on redirect

### LINKED ISSUE #1711: redirecting >11 times causes "MaxListenersExceededWarning"
## Bug Description

Each redirect adds one `terminated` listener to fetchParams and it's never removed.

## Reproducible By

```mjs
import { once } from 'events'
import { createServer } from 'http'
import { fetch } from 'undici'

let redirects = 0

const server = createServer((req, res) => {
  if (redirects === 15) {
    res.end('Okay goodbye')
    return
  }

  res.writeHead(302, {
    Location: `/${redirects++}`
  })
  res.end()
}).listen(0)

await once(server, 'listening')

const url = `http://localhost:${server.address().port}`

fetch(url, { redirect: 'follow' })
```

## Logs & Screenshots

```
MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 terminated listeners added to [Fetch]. Use emitter.setMaxListeners() to increase limit
```

### PR DESCRIPTION
**Pull request by KhafraDev**
Disclaimer: this is a pretty terrible solution, but I couldn't find another way of fixing this.

Fixes #1711 

- The listener gets added here
https://github.com/nodejs/undici/blob/7102eb0b7f544cee7669aaf06eb94f38b22aba86/lib/fetch/index.js#L1813
- This is where it should be removed (& onError). If we remove it here it doesn't work, my best guess is that it gets gc'ed? I can't move `onAborted` outside of this scope as it relies on `response`, and if we use a closure (etc.) the same problem would happen.
https://github.com/nodejs/undici/blob/7102eb0b7f544cee7669aaf06eb94f38b22aba86/lib/fetch/index.js#L2028

If you have any idea of how to fix this, please let me know. I spent maybe 45 minutes trying anything I could think of.

Also not sure how I would add tests for this.

### CODE REVIEW

**Comment by mcollina** [comment]
Could you add a unit test?

**Comment by mcollina** [comment]
The code looks good. I would remove the individual listener and not all.

**Comment by ronag** [comment]
For now... I think it's easier just to increase the listener limit, i.e. `setMaxListeners`.

File: lib/fetch/index.js [hunk 1]
```diff
@@ -73,6 +73,11 @@ class Fetch extends EE {
     this.connection = null
     this.dump = false
     this.state = 'ongoing'
+    // 2 terminated listeners get added per request,
+    // but only 1 gets removed. If there are 20 redirects,
+    // 21 listeners will be added.
+    // See https://github.com/nodejs/undici/issues/1711
+    this.setMaxListeners(21)
```
Comment (ronag) [inline lib/fetch/index.js:80]:
```suggestion
    // TODO (fix): Find and fix root cause for leaked listener.
    this.setMaxListeners(21)
```

**Comment by jonseymour** [comment]
Did this PR actually fix anything or did it just mask the problem so that an annoying message got printed less frequently?

**Comment by KhafraDev** [comment]
Masked the issue.

**Comment by KhafraDev** [comment]
I'll have a PR that actually fixes it, wasn't that complicated.

**Comment by jonseymour** [comment]
Could I clarify, Is there is an actual issue with fetch listeners being added but not removed in the presence of redirect responses?

**Comment by Uzlopak** [comment]
its not like you have a memory leak.

**Comment by jonseymour** [comment]
ok, so the emission of the messages was simply caused by bursty acquisition of listeners prior to GC occurring?

**Comment by Uzlopak** [comment]
Basically. An EventEmitter could have hundreds of listeners for a single event without being an issue. But the case is very rare, so the node core devs decided to warn the devs if they add more than 10 listeners to an event without maybe actively noticing it, as it indicates a bug, which could slow down the performance, without being that obvious while you are developing.

**Pull request merged**